TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Workspace.php

```diff
@@ -403,62 +402,83 @@ public function publishNodes(array $nodes, Workspace $targetWorkspace)
      *
      * The specified workspace must be a base workspace of this workspace.
      *
-     * @param NodeInterface $node The node to publish
+     * @param NodeInterface $nodeToPublish The node to publish
      * @param Workspace $targetWorkspace The workspace to publish to
      * @return void
      * @api
      */
-    public function publishNode(NodeInterface $node, Workspace $targetWorkspace)
+    public function publishNode(NodeInterface $nodeToPublish, Workspace $targetWorkspace)
     {
-        if ($this->baseWorkspace === null) {
+        if ($this->publishNodeCanBeSkipped($nodeToPublish, $targetWorkspace)) {
             return;
         }
+        $this->emitBeforeNodePublishing($nodeToPublish, $targetWorkspace);
+
+        $correspondingNodeDataInTargetWorkspace = $this->findCorrespondingNodeDataInTargetWorkspace($nodeToPublish, $targetWorkspace);
+        $matchingNodeVariantExistsInTargetWorkspace = ($correspondingNodeDataInTargetWorkspace !== null && $correspondingNodeDataInTargetWorkspace->getDimensionValues() === $nodeToPublish->getDimensions());
+
+        // Save the original node workspace because the workspace of $nodeToPublish can be changed by replaceNodeData():
+        $originalNodeWorkspace = $nodeToPublish->getWorkspace();
+
+        if ($matchingNodeVariantExistsInTargetWorkspace) {
+            $this->replaceNodeData($nodeToPublish, $correspondingNodeDataInTargetWorkspace);
+            $this->moveNodeVariantsInOtherWorkspaces($nodeToPublish->getIdentifier(), $nodeToPublish->getPath(), $originalNodeWorkspace, $targetWorkspace);
+        } else {
+            $this->moveNodeVariantToTargetWorkspace($nodeToPublish, $targetWorkspace);
+        }
+
+        $this->emitAfterNodePublishing($nodeToPublish, $targetWorkspace);
+    }
+
+    /**
+     * Checks if the given node can / needs to be published to the given target workspace or if that operation can
+     * be skipped.
+     *
+     * @param NodeInterface $node The node to be published
+     * @param Workspace $targetWorkspace The target workspace
+     * @return bool
+     */
+    protected function publishNodeCanBeSkipped(NodeInterface $node, Workspace $targetWorkspace)
+    {
+        if ($this->baseWorkspace === null) {
+            return true;
+        }
         if ($node->getWorkspace() !== $this) {
-            return;
+            return true;
         }
         // Might happen if a node which has been published during an earlier call of publishNode() is attempted to
         // be published again:
         if ($node->getWorkspace() === $targetWorkspace) {
```

Comment:
Shouldn't this condition be unnecessary anyway after the previous one? It would mean `$this` must be the `$targetWorkspace` which seems strange, but maybe that edge case exists?